File: Kotlin/ComposeMultiplatform/src/main/kotlin/ui/components/TodoList.kt
package ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import data.TabFilter
import data.Todo
import ui.Theme

@Composable
fun TodoList(
    todos: List<Todo>,
    activeTab: TabFilter,
    onToggle: (Int) -> Unit,
) {
    val filtered = when (activeTab) {
        TabFilter.All       -> todos
        TabFilter.Active    -> todos.filter { !it.isCompleted }
        TabFilter.Completed -> todos.filter { it.isCompleted }
    }

    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        verticalArrangement = Arrangement.spacedBy(Theme.ITEM_MB),
    ) {
        items(filtered, key = { it.id }) { todo ->
            TodoItem(todo = todo, onToggle = { onToggle(todo.id) })
        }
    }
}

@Composable
private fun TodoItem(todo: Todo, onToggle: () -> Unit) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .height(Theme.ITEM_H)
            .clip(RoundedCornerShape(Theme.RADIUS))
            .background(Theme.SURFACE)
            .padding(Theme.ITEM_PAD),
    ) {
        // Circular checkbox
        Canvas(
            modifier = Modifier
                .size(Theme.CB_SIZE)
                .clickable(
                    indication = null,
                    interactionSource = remember { MutableInteractionSource() },
                    onClick = onToggle,
                ),
        ) {
            val r = size.minDimension / 2f
            val center = Offset(r, r)
            if (todo.isCompleted) {
                drawCircle(
                    color = Theme.CB_BORDER,
                    radius = r - 1.dp.toPx(),
                    style = Stroke(width = 2.dp.toPx()),
                )
                // Checkmark in accent color
                val sw = 1.5.dp.toPx()
                drawLine(
                    color = Theme.ACCENT,
                    start = Offset(center.x - r * 0.45f, center.y),
                    end   = Offset(center.x - r * 0.1f,  center.y + r * 0.45f),
                    strokeWidth = sw,
                )
                drawLine(
                    color = Theme.ACCENT,
                    start = Offset(center.x - r * 0.1f, center.y + r * 0.45f),
                    end   = Offset(center.x + r * 0.5f, center.y - r * 0.4f),
                    strokeWidth = sw,
                )
            } else {
                drawCircle(
                    color = Theme.CB_BORDER,
                    radius = r - 1.dp.toPx(),
                    style = Stroke(width = 2.dp.toPx()),
                )
            }
        }

        Spacer(modifier = Modifier.width(Theme.CB_MR))

        Text(
            text = todo.task,
            color = if (todo.isCompleted) Theme.MUTED else Theme.TEXT,
            fontSize = Theme.FONT_NORM,
            style = TextStyle(
                textDecoration = if (todo.isCompleted) TextDecoration.LineThrough else TextDecoration.None,
                fontSize = Theme.FONT_NORM,
                color = if (todo.isCompleted) Theme.MUTED else Theme.TEXT,
            ),
        )
    }
}

File: Python/Flet/components/todo_list.py
import flet as ft
from typing import Callable, List
from data import Todo
from theme import Theme


def build_checkbox(completed: bool) -> ft.Container:
    if completed:
        return ft.Container(
            content=ft.Text("✓", size=10, color=Theme.ACCENT),
            width=16,
            height=16,
            border_radius=8,
            border=ft.Border.all(2, Theme.CB_BDR),
            alignment=ft.Alignment(0, 0),
        )
    else:
        return ft.Container(
            width=16,
            height=16,
            border_radius=8,
            border=ft.Border.all(2, Theme.CB_BDR),
        )


def build_todo_item(todo: Todo, on_toggle: Callable[[int], None]) -> ft.Container:
    def handle_toggle(e):
        on_toggle(todo.id)

    checkbox = build_checkbox(todo.is_completed)

    label = ft.Text(
        todo.task,
        size=14,
        color=Theme.MUTED if todo.is_completed else Theme.TEXT,
        style=ft.TextStyle(
            decoration=ft.TextDecoration.LINE_THROUGH if todo.is_completed else ft.TextDecoration.NONE,
        ),
    )

    row = ft.Row(
        controls=[
            ft.GestureDetector(
                content=checkbox,
                on_tap=handle_toggle,
                mouse_cursor=ft.MouseCursor.CLICK,
            ),
            ft.Container(width=18),
            label,
        ],
        spacing=0,
        vertical_alignment=ft.CrossAxisAlignment.CENTER,
    )

    return ft.Container(
        content=row,
        bgcolor=Theme.SURFACE,
        border_radius=4,
        padding=10,
        height=35,
        margin=ft.Margin(0, 0, 0, 5),
    )


def build_todo_list(
    todos: List[Todo],
    on_toggle: Callable[[int], None],
) -> ft.Column:
    items = [build_todo_item(t, on_toggle) for t in todos]
    return ft.Column(
        controls=items,
        spacing=0,
        scroll=ft.ScrollMode.AUTO,
        expand=True,
    )